policy-controller/k8s/index/src/outbound/index.rs

```diff
@@ -370,6 +363,23 @@ fn is_parent_service(parent: &ParentReference) -> bool {
         .unwrap_or(false)
 }
 
+#[inline]
```

Comment:
This is me asking for purely academic reasons: `inline` here basically means the function is hoisted up into its callsite when we translate this to machine code as opposed to living somewhere else in memory?

Reply:
Yeah, that's basically what inlining means. Technically, the `#[inline]` attribute is just a _hint_ to the compiler that means it _might_ want to inline the function --- `#[inline(always)]` actually *guarantees* that a function will be inlined.

Inlining may have both performance benefits and performance costs, depending on the particular code in question. Inlining short functions avoids pushing a stack frame for small amounts of code, which can have a significant benefit, but it also increases code size , resulting in a larger binary, and also potentially impacts cache performance. Generally, compilers use a complex set of heuristics to determine whether or not to inline a particular function call, which is why the `#[inline]` attribute is just a _hint_ that is considered as part of those heuristics. This means that it's safe to use `#[inline]` fairly liberally, while `#[inline(always)]` should basically only be used in _very_ specific situations[^1] which we generally don't encounter in Linkerd.

One other note is that the `#[inline]` attribute enables functions to be inlined across crate boundaries, which Rust generally doesn't do.

[^1]: e.g. in embedded systems or very low-level programming like kernel modules, when there may be a hard limit on stack depth.